File: php/services/CurrencyConfig.php
<?php

namespace Services;

class CurrencyConfig
{
    private static $currencies = [
        'USD' => [
            'code' => 'USD',
            'name' => 'US Dollar',
            'symbol' => '$',
            'symbolPosition' => 'before',
            'decimals' => 2,
            'decimalSeparator' => '.',
            'thousandsSeparator' => ',',
            'country' => 'US',
            'flag' => '🇺🇸'
        ],
        'EUR' => [
            'code' => 'EUR',
            'name' => 'Euro',
            'symbol' => '€',
            'symbolPosition' => 'after',
            'decimals' => 2,
            'decimalSeparator' => ',',
            'thousandsSeparator' => '.',
            'country' => 'GB', // GP API uses GB for EUR processing
            'flag' => '🇪🇺'
        ],
        'GBP' => [
            'code' => 'GBP',
            'name' => 'British Pound',
            'symbol' => '£',
            'symbolPosition' => 'before',
            'decimals' => 2,
            'decimalSeparator' => '.',
            'thousandsSeparator' => ',',
            'country' => 'GB',
            'flag' => '🇬🇧'
        ],
        'CAD' => [
            'code' => 'CAD',
            'name' => 'Canadian Dollar',
            'symbol' => 'C$',
            'symbolPosition' => 'before',
            'decimals' => 2,
            'decimalSeparator' => '.',
            'thousandsSeparator' => ',',
            'country' => 'CA',
            'flag' => '🇨🇦'
        ],
        'AUD' => [
            'code' => 'AUD',
            'name' => 'Australian Dollar',
            'symbol' => 'A$',
            'symbolPosition' => 'before',
            'decimals' => 2,
            'decimalSeparator' => '.',
            'thousandsSeparator' => ',',
            'country' => 'AU',
            'flag' => '🇦🇺'
        ],
        'JPY' => [
            'code' => 'JPY',
            'name' => 'Japanese Yen',
            'symbol' => '¥',
            'symbolPosition' => 'before',
            'decimals' => 0,
            'decimalSeparator' => '',
            'thousandsSeparator' => ',',
            'country' => 'JP',
            'flag' => '🇯🇵'
        ]
    ];

    /**
     * Get all supported currencies
     *
     * @return array
     */
    public static function getAllCurrencies(): array
    {
        return self::$currencies;
    }

    /**
     * Get currency configuration by code
     *
     * @param string $code Currency code (USD, EUR, etc.)
     * @return array|null Currency configuration or null if not found
     */
    public static function getCurrency(string $code): ?array
    {
        $code = strtoupper($code);
        return self::$currencies[$code] ?? null;
    }

    /**
     * Check if currency is supported
     *
     * @param string $code Currency code
     * @return bool
     */
    public static function isSupported(string $code): bool
    {
        return isset(self::$currencies[strtoupper($code)]);
    }

    /**
     * Get country code for currency (for GP API configuration)
     *
     * @param string $currencyCode Currency code
     * @return string Country code
     */
    public static function getCountryCode(string $currencyCode): string
    {
        $currency = self::getCurrency($currencyCode);
        return $currency ? $currency['country'] : 'US';
    }

    /**
     * Get number of decimal places for currency
     *
     * @param string $currencyCode Currency code
     * @return int Number of decimals
     */
    public static function getDecimals(string $currencyCode): int
    {
        $currency = self::getCurrency($currencyCode);
        return $currency ? $currency['decimals'] : 2;
    }

    /**
     * Format amount according to currency rules
     *
     * @param float $amount Amount to format
     * @param string $currencyCode Currency code
     * @return string Formatted amount with currency symbol
     */
    public static function formatAmount(float $amount, string $currencyCode): string
    {
        $currency = self::getCurrency($currencyCode);
        if (!$currency) {
            return number_format($amount, 2);
        }

        $formatted = number_format(
            $amount,
            $currency['decimals'],
            $currency['decimalSeparator'],
            $currency['thousandsSeparator']
        );

        if ($currency['symbolPosition'] === 'before') {
            return $currency['symbol'] . $formatted;
        } else {
            return $formatted . ' ' . $currency['symbol'];
        }
    }

    /**
     * Get supported currency codes
     *
     * @return array
     */
    public static function getSupportedCodes(): array
    {
        return array_keys(self::$currencies);
    }

    /**
     * Validate and normalize currency code
     *
     * @param string|null $code Currency code
     * @return string Valid currency code (defaults to USD)
     */
    public static function validateCurrency(?string $code): string
    {
        if (!$code) {
            return 'USD';
        }

        $code = strtoupper(trim($code));
        return self::isSupported($code) ? $code : 'USD';
    }
}

File: java/src/main/java/com/globalpayments/example/services/CurrencyConfig.java
package com.globalpayments.example.services;

import java.text.NumberFormat;
import java.util.*;

/**
 * Currency Configuration
 * Manages currency metadata and formatting rules
 */
public class CurrencyConfig {

    public static class Currency {
        public String code;
        public String name;
        public String symbol;
        public String symbolPosition;
        public int decimals;
        public String decimalSeparator;
        public String thousandsSeparator;
        public String country;
        public String flag;

        public Currency(String code, String name, String symbol, String symbolPosition,
                       int decimals, String decimalSeparator, String thousandsSeparator,
                       String country, String flag) {
            this.code = code;
            this.name = name;
            this.symbol = symbol;
            this.symbolPosition = symbolPosition;
            this.decimals = decimals;
            this.decimalSeparator = decimalSeparator;
            this.thousandsSeparator = thousandsSeparator;
            this.country = country;
            this.flag = flag;
        }
    }

    private static final Map<String, Currency> currencies = new HashMap<>();

    static {
        currencies.put("USD", new Currency(
            "USD", "US Dollar", "$", "before",
            2, ".", ",", "US", "🇺🇸"
        ));
        currencies.put("EUR", new Currency(
            "EUR", "Euro", "€", "after",
            2, ",", ".", "GB", "🇪🇺"  // GP API uses GB for EUR processing
        ));
        currencies.put("GBP", new Currency(
            "GBP", "British Pound", "£", "before",
            2, ".", ",", "GB", "🇬🇧"
        ));
        currencies.put("CAD", new Currency(
            "CAD", "Canadian Dollar", "C$", "before",
            2, ".", ",", "CA", "🇨🇦"
        ));
        currencies.put("AUD", new Currency(
            "AUD", "Australian Dollar", "A$", "before",
            2, ".", ",", "AU", "🇦🇺"
        ));
        currencies.put("JPY", new Currency(
            "JPY", "Japanese Yen", "¥", "before",
            0, "", ",", "JP", "🇯🇵"
        ));
    }

    /**
     * Get all supported currencies
     */
    public static Map<String, Currency> getAllCurrencies() {
        return new HashMap<>(currencies);
    }

    /**
     * Get currency configuration by code
     */
    public static Currency getCurrency(String code) {
        return currencies.get(code.toUpperCase());
    }

    /**
     * Check if currency is supported
     */
    public static boolean isSupported(String code) {
        return currencies.containsKey(code.toUpperCase());
    }

    /**
     * Get country code for currency (for GP API configuration)
     */
    public static String getCountryCode(String currencyCode) {
        Currency currency = getCurrency(currencyCode);
        return currency != null ? currency.country : "US";
    }

    /**
     * Format amount according to currency rules
     */
    public static String formatAmount(double amount, String currencyCode) {
        Currency currency = getCurrency(currencyCode);
        if (currency == null) {
            return String.format("%.2f", amount);
        }

        // Format with proper decimal places
        String formatted = String.format("%." + currency.decimals + "f", amount);

        // Replace decimal separator
        if (!currency.decimalSeparator.isEmpty()) {
            formatted = formatted.replace(".", currency.decimalSeparator);
        }

        // Add thousands separator
        String[] parts = formatted.split("[,.]");
        if (parts.length > 0) {
            String integerPart = parts[0];
            StringBuilder result = new StringBuilder();
            int count = 0;
            for (int i = integerPart.length() - 1; i >= 0; i--) {
                if (count > 0 && count % 3 == 0) {
                    result.insert(0, currency.thousandsSeparator);
                }
                result.insert(0, integerPart.charAt(i));
                count++;
            }

            if (currency.decimals > 0 && parts.length > 1) {
                result.append(currency.decimalSeparator).append(parts[1]);
            }
            formatted = result.toString();
        }

        // Add currency symbol
        if ("before".equals(currency.symbolPosition)) {
            return currency.symbol + formatted;
        } else {
            return formatted + " " + currency.symbol;
        }
    }

    /**
     * Get supported currency codes
     */
    public static Set<String> getSupportedCodes() {
        return currencies.keySet();
    }

    /**
     * Validate and normalize currency code
     */
    public static String validateCurrency(String code) {
        if (code == null || code.trim().isEmpty()) {
            return "USD";
        }
        String upperCode = code.toUpperCase().trim();
        return isSupported(upperCode) ? upperCode : "USD";
    }
}
